### ISSUE TITLE
unexpected transition happened with router change after calling applicationRef.attachView 

### ISSUE THREAD

**Issue by vthinkxie** [angular/angular#34718]
# 🐞 bug report

### Affected Package

`@angular/core`
`@angular/router`

### Is this a regression?
no

### Description

The unexpected transition will happen when router change after attaching the component with the `applicationRef.attachView`

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-router-transition-bug?file=src/app/app.component.ts

1. click the router link, you would find the unexpected transition happened
2. comment this line `this.applicationRef.attachView(view.hostView);`
3. click the router link again, you would find the transition won't happen anymore.

## 🔥 Exception or Error
![Kapture 2020-01-10 at 15 32 20](https://user-images.githubusercontent.com/1506722/72134234-72243600-33be-11ea-8249-687619966abe.gif)

## 🌍  Your Environment

**Angular Version:**
<pre><code>

@angular-devkit/architect         0.803.21
@angular-devkit/build-angular     0.803.21
@angular-devkit/build-optimizer   0.803.21
@angular-devkit/build-webpack     0.803.21
@angular-devkit/core              8.3.21
@angular-devkit/schematics        8.3.21
@angular/cdk                      8.2.3
@angular/cli                      8.3.21
@ngtools/webpack                  8.3.21
@schematics/angular               8.3.21
@schematics/update                0.803.21
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2

</code></pre>

**Anything else relevant?**

the bug still exists in the latest angular version 9.0.0-rc.8

It breaks all transition style in our components after any overlay created with `@angular/cdk`

here is one ng-zorro-antd bug caused by this issue

https://stackblitz.com/edit/ng-zorro-antd-start-qgr95a?file=src%2Fapp%2Fhome%2Fhome.component.html

hope it could be fixed before angular 9 release.
ref https://github.com/angular/components/issues/14646

**Comment by atscott** [comment]
It appears this behaves as expected based on stackblitz reproduction. `componentFactory.create` is creating a new root outside of the existing tree. You need to attach it with `applicationRef.attachView` for it to be [change detected](https://angular.io/api/core/ApplicationRef#attachview). If you comment that line, then the component is not change detected and host bindings are not evaluated.

If you still feel there is an issue, please provide further details in a new issue.

**Comment by vthinkxie** [comment]
Hi @atscott 
the host className was not binding to the attached component.
I don't think it is reasonable that attaching a component would affect other component's host binding className transition behavior.

**Comment by atscott** [comment]
Sorry, I misunderstood the example and was looking at the wrong thing. What would you expect the behavior to be here? It seems like the `attachView` call actually causes the transition to happen when it should not. The host binding class `.transition` is added immediately at the time of creation so the ease transition should not happen. See [this](https://stackoverflow.com/questions/24148403/trigger-css-transition-on-appended-element) related question on stack overflow), specifically 
> When element is added, reflow is needed. The same applies to adding the class. However when you do both in single javascript round, browser takes its chance to optimize out the first one. In that case, there is only single (initial and final at the same time) style value, so no transition is going to happen.

**Comment by vthinkxie** [comment]
Hi @atscott 
I would expect the className behavior the same no matter if there is any view attached to the app component.

I have debugged this issue with DOM breakpoint and find that
1. the transition class was added immediately after router change if there is **no attached view**
<img width="1465" alt="without-attach" src="https://user-images.githubusercontent.com/1506722/72401103-183cbb00-3786-11ea-8e38-2081dd137f27.png">

2. the transition class was not added after router change if **there is an attached view**
<img width="1467" alt="after-attched" src="https://user-images.githubusercontent.com/1506722/72401135-28ed3100-3786-11ea-8ada-a2d770c8a6f8.png">
and the transition classname added after the next tick, which caused the unexpected transtion.

**Comment by atscott** [comment]
Yes, I agree that it should be consistent. For what it's worth, I can't reproduce this unexpected transition when running the same example with Ivy enabled. Can you try running your app with the latest v9 RC to see if that resolves your issue?

**Comment by vthinkxie** [comment]
Yes, I can reproduce it with the latest v9 RC, I am creating a repo now, wait for a moment

**Comment by vthinkxie** [comment]
@atscott 
here is the repo https://github.com/vthinkxie/ng-transition-bug

**Comment by vthinkxie** [comment]
Hi @atscott 
I think the bug was caused by the instantiateRendererFactory here, it would work fine if comment these lines below.
https://github.com/angular/angular/blob/bf8ba89394ffd6d2047261e89d9f41e6ef66be3e/packages/platform-browser/animations/src/providers.ts#L48-L52
I hope this would help.

**Comment by vthinkxie** [comment]
Hi @pkozlowski-opensource 
Would this issue be fixed in the 9.1.0 version?
Our component libs depend on the angular animation heavily, we implement a hack directive to skip this issue, but still hopes angular could fix this issue soon
https://github.com/NG-ZORRO/ng-zorro-antd/blob/af4051eb11a5c9a571d7ccdea75774ea786ba990/components/core/transition-patch/transition-patch.directive.ts#L18

**Comment by atscott** [comment]
TL;DR - Jump to the bottom of this comment for possible solutions.

There's still a fair amount I don't know, but here's what I understand so far:

* Part of this is similar to https://github.com/angular/angular/issues/37279#issuecomment-715618229
* How this example relates to the point above is that the router outlet creates the component, then calls `markForCheck` router. Meaning that the component is created but change detection has not run yet (dynamic host bindings are assigned during change detection, not during element creation)
* Because you attached the viewRef in the constructor of the `AppComponent`, it gets added to the `ApplicationRef`'s `_views` _before_ `AppComponent`. This is significant because that means when change detection runs, it will be processed first when change detection runs. After each root view component, the animation engine flushes. See 
https://github.com/angular/angular/blob/90c0772743fda18e6b4bab6feaa731038b872cbc/packages/core/src/render3/instructions/shared.ts#L1876-L1880 ->  https://github.com/angular/angular/blob/90c0772743fda18e6b4bab6feaa731038b872cbc/packages/core/src/render3/instructions/shared.ts#L492 -> https://github.com/angular/angular/blob/90c0772743fda18e6b4bab6feaa731038b872cbc/packages/platform-browser/animations/src/animation_renderer.ts#L122. I _think_ this flush results in the adding of the component happening in a "separate tick" from the hostbinding class being added (again, similar to the first bullet point)
  * Note that this isn't necessarily specific to the `attachView` call. There are other things that might cause the browser to "see" the addition of the class with the transition as a separate operation (again, see the first bullet point)

There are a couple possible solutions for you:

1. Preferred solution: change your host binding to be static. Static host bindings are applied immediately on creation of the component. I highly recommend this option for all static host bindings to avoid unexpected behavior like this instead of using `'[class.X]': "true"`. So convert `host: {"[class.transition]": "true"}` to `host: {"class": "transition"}`.
2. If solution #1 is not an option for you because your host binding really is static, maybe you can call `applicationRef.attachView` in the `ngOnInit` of `ApplicationRef`. This will mean that the root view with the router outlet is processed first
3. (a tiny bit messy because it involves casting your forked class to a `RouterOutlet` to pass to `parentContexts.onChildOutletCreated`): Fork `router-outlet`  and in `activateWith`, call `this.changeDetector.detectChanges();` instead of or right before `this.changeDetector.markForCheck();`. This will cause the change detection to run for your created component immediately and the host binding should be applied before the browser can "see" that the class wasn't there at the start.


Given that there are several possible solutions here, I'm setting the priority lower. I don't think there's much that can be done in the core package to fix this. Dynamic bindings _have_ to be processed later because they could depend on component data that in turn depends on change detection running. This could be addressed in the `router-outlet` by calling `detectChanges` in `activateWith` like option #3, but I'm wary of making this change because it might cause other unexpected failures (changing the behavior of CD in the core components is quite often a breaking change).

**Comment by vthinkxie** [comment]
thanks a lot @atscott 
we will try your solutions

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
